File: Squealer/SquealerConsole/VersionCheck.vb
Public Class VersionCheck

    Private Const VersionInfoFilename As String = "version.xml"
    Private Const DownloadFilename As String = "Squealer{0}-Install.zip"

    Private Class Metadata

#Region " Properties "

        Private _Version As Version
        Public Property Version As Version
            Get
                Return _Version
            End Get
            Set(value As Version)
                _Version = value
            End Set
        End Property

        Public ReadOnly Property IsUpdateAvailable As Boolean
            Get
                Return My.Application.Info.Version.CompareTo(_Version) < 0
            End Get
        End Property

        Private _Updated As DateTime
        Public Property Updated As DateTime
            Get
                Return _Updated
            End Get
            Set(value As DateTime)
                _Updated = value
            End Set
        End Property

        Private _Hidden As Boolean
        Public Property Hidden As Boolean
            Get
                Return _Hidden
            End Get
            Set(value As Boolean)
                _Hidden = value
            End Set
        End Property

        Private _About As String
        Public Property About As String
            Get
                Return _About
            End Get
            Set(value As String)
                _About = value
            End Set
        End Property

        Private _ZipFile As String
        Public Property ZipFile As String
            Get
                Return _ZipFile
            End Get
            Set(value As String)
                _ZipFile = value
            End Set
        End Property

#End Region

        Public Sub New()
        End Sub

        Public Sub New(ver As Version, updated As DateTime, hidden As Boolean, about As String, sourceurl As String)
            Me._Version = ver
            Me._Updated = updated
            Me._Hidden = hidden
            Me._About = about
            Me._ZipFile = String.Format(sourceurl & DownloadFilename, _Version.ToString.Replace(".", ""), sourceurl)
        End Sub

    End Class

#Region " Output "

    Public Sub CreateMetadata(sourceurl As String)
        Dim info As New Metadata(My.Application.Info.Version, DateTime.UtcNow, False, WhatsNew, sourceurl)
        Dim f As New TempFileHandler(".xml")
        Generate(info).Save(f.Filename)
        f.Show()
    End Sub

    Private Function Generate(ByRef info As Metadata) As Xml.XmlDocument

        Dim doc As New Xml.XmlDocument
        doc.AppendChild(doc.CreateXmlDeclaration("1.0", "us-ascii", Nothing))
        Dim squealer As Xml.XmlElement = doc.CreateElement("Squealer")
        doc.AppendChild(squealer)
        squealer.SetAttribute(NameOf(info.Version), info.Version.ToString)
        squealer.SetAttribute(NameOf(info.Updated), info.Updated.ToString)
        squealer.SetAttribute(NameOf(info.Hidden), info.Hidden.ToString)
        squealer.SetAttribute(NameOf(info.ZipFile), info.ZipFile)
        Dim about As Xml.XmlElement = doc.CreateElement(NameOf(info.About))
        squealer.AppendChild(about)
        Dim abouttext As Xml.XmlCDataSection = doc.CreateCDataSection("")
        about.AppendChild(abouttext)
        abouttext.InnerText = info.About
        Generate = doc

    End Function

    Public Shared Function WhatsNew() As String
        Return My.Resources.ChangeLog.Remove(My.Resources.ChangeLog.IndexOf("^break")).Replace(My.Application.Info.Version.ToString, String.Empty).Trim
    End Function
    Public Shared Function ChangeLog() As String
        Return My.Resources.ChangeLog.Replace("^break", String.Empty).Trim
    End Function

#End Region

#Region " Input "

    Public Sub DisplayVersionCheckResults(sourceurl As String, ismediadefaultlocation As Boolean)

        Dim m As Metadata = DownloadMetadata(sourceurl)
        If m.IsUpdateAvailable Then
            Textify.SayBulletLine(Textify.eBullet.Hash, "A new version is available.", New Textify.ColorScheme(ConsoleColor.Yellow))
            Textify.SayBulletLine(Textify.eBullet.Hash, String.Format("Squealer {0} released on {1}.", m.Version, m.Updated.ToShortDateString))
            Console.WriteLine()
            Console.WriteLine(m.About)
            Console.WriteLine()
            Textify.SayBulletLine(Textify.eBullet.Hash, String.Format("Get the latest version at {0}", IIf(ismediadefaultlocation, Constants.HomePage, m.ZipFile)))
        Else
            Textify.SayBulletLine(Textify.eBullet.Hash, String.Format("You have the latest version of Squealer ({0}).", My.Application.Info.Version), New Textify.ColorScheme(ConsoleColor.White))
        End If

    End Sub

    Private Function DownloadMetadata(sourceurl As String) As Metadata

        Dim d As New Metadata()
        Dim client As New Net.WebClient
        Dim x As New Xml.XmlDocument
        Dim r As New IO.StreamReader(client.OpenRead(sourceurl & VersionInfoFilename))
        x.LoadXml(r.ReadToEnd)

        Try
            Dim Node As Xml.XmlNode = x.SelectSingleNode("/Squealer")
            d.Version = Version.Parse(Node.Attributes(NameOf(d.Version)).Value)
            d.Updated = DateTime.Parse(Node.Attributes(NameOf(d.Updated)).Value)
            d.Hidden = Boolean.Parse(Node.Attributes(NameOf(d.Hidden)).Value)
            d.ZipFile = Node.Attributes(NameOf(d.ZipFile)).Value.ToString
            Node = x.SelectSingleNode("/Squealer/About")
            d.About = x.InnerText
            Return d
        Catch ex As Exception
            Return Nothing
        End Try

        '<?xml version="1.0" encoding="us-ascii"?>
        '<Squealer Version = "1.2.3.45" Updated="mm/dd/yyyy 12:00:00 AM" Hidden="False" Filename="https://s3_xxx/Squealerabcde12345-Install.zip">
        '  <About><![CDATA[Release notes]]></About>
        '</Squealer>

    End Function

    Public Sub DownloadLatestInstaller(sourceurl As String)

        Dim d As Metadata = DownloadMetadata(sourceurl)
        Dim dest As String = Environment.GetFolderPath(Environment.SpecialFolder.MyDocuments) & "\" & System.IO.Path.GetFileName(d.ZipFile)
        Textify.Write("Downloading ", New Textify.ColorScheme(ConsoleColor.Green))
        Textify.Write(d.ZipFile, New Textify.ColorScheme(ConsoleColor.DarkYellow))
        Textify.Write(" to ", New Textify.ColorScheme(ConsoleColor.Green))
        Textify.WriteLine(dest, New Textify.ColorScheme(ConsoleColor.White))
        Console.WriteLine()

        Using client As New System.Net.WebClient()
            client.DownloadFile(d.ZipFile, dest)
        End Using

    End Sub

    Public Sub DownloadLatestEzBinary(sourceurl As String, destfile As String)
        ' silently get the latest script
        Try
            Using client As New System.Net.WebClient()
                client.DownloadFile(sourceurl, destfile)
            End Using
        Catch ex As Exception
        End Try
    End Sub

#End Region

End Class

File: Squealer/SquealerConsoleCSharp/VersionCheck.cs
using Spectre.Console;
using System.Reflection;
using System.Xml;

namespace SquealerConsoleCSharp
{
    internal class VersionCheck
    {
        private const string VersionInfoFilename = "version.xml";

        private class Metadata
        {
            public Version? Version { get; set; }
            public DateTime Updated { get; set; }
            public bool Hidden { get; set; }
            public string About { get; set; } = string.Empty;
            public string ZipFile { get; set; } = string.Empty;

            public bool IsUpdateAvailable
            {
                get
                {
                    var current = Assembly.GetEntryAssembly()?.GetName().Version;
                    return current != null && Version != null && current.CompareTo(Version) < 0;
                }
            }
        }

        public void DisplayVersionCheckResults(string sourceUrl)
        {
            var m = DownloadMetadata(sourceUrl);
            if (m.Version == null) return;

            if (m.IsUpdateAvailable)
            {
                AnsiConsole.MarkupLine("[yellow]# A new version is available.[/]");
                AnsiConsole.MarkupLine($"# Squealer {m.Version} released on {m.Updated.ToShortDateString()}.");
                AnsiConsole.WriteLine();
                AnsiConsole.WriteLine(m.About);
                AnsiConsole.WriteLine();
                AnsiConsole.MarkupLine($"# Get the latest version at {Constants.HomePage}");
            }
            else
            {
                var ver = Assembly.GetEntryAssembly()?.GetName().Version;
                AnsiConsole.MarkupLine($"[white]# You have the latest version of Squealer ({ver}).[/]");
            }
        }

        public void DownloadLatestInstaller(string sourceUrl)
        {
            var m = DownloadMetadata(sourceUrl);
            if (string.IsNullOrEmpty(m.ZipFile)) return;

            var dest = Path.Combine(
                Environment.GetFolderPath(Environment.SpecialFolder.MyDocuments),
                Path.GetFileName(m.ZipFile));

            AnsiConsole.MarkupLine($"[green]Downloading[/] [darkyellow]{m.ZipFile}[/] [green]to[/] [white]{dest}[/]");
            AnsiConsole.WriteLine();

            using var client = new HttpClient();
            var data = client.GetByteArrayAsync(m.ZipFile).GetAwaiter().GetResult();
            File.WriteAllBytes(dest, data);
        }

        private static Metadata DownloadMetadata(string sourceUrl)
        {
            var m = new Metadata();
            try
            {
                using var client = new HttpClient();
                client.Timeout = TimeSpan.FromSeconds(10);
                var xml = client.GetStringAsync(sourceUrl + VersionInfoFilename).GetAwaiter().GetResult();

                var doc = new XmlDocument();
                doc.LoadXml(xml);
                var node = doc.SelectSingleNode("/Squealer");
                if (node?.Attributes == null) return m;

                m.Version = Version.Parse(node.Attributes["Version"]!.Value);
                m.Updated = DateTime.Parse(node.Attributes["Updated"]!.Value);
                m.Hidden = bool.Parse(node.Attributes["Hidden"]!.Value);
                m.ZipFile = node.Attributes["ZipFile"]?.Value ?? string.Empty;
                m.About = node.SelectSingleNode("About")?.InnerText ?? string.Empty;
            }
            catch
            {
                AnsiConsole.MarkupLine("[red]Could not check for updates. Check your connection and try again.[/]");
            }
            return m;
        }
    }
}
